### PR TITLE
[mono][jit] Don't truncate the high bits when ldelema index is nint

### PR DESCRIPTION
**Pull request by BrzVlad**
JIT only fix from https://github.com/dotnet/runtime/pull/73799. Interp fix might be more invasive.

### CODE REVIEW

**Comment by BrzVlad** [comment]
A little bit of complication because, according to comments, emitting the SEXT_I4 normally would interfere with bounds check removal. @vargaz Please take a look

**Comment by BrzVlad** [comment]
@vargaz Could I have a review on this ?

File: src/mono/mono/mini/method-to-ir.c [hunk 1]
```diff
@@ -4444,8 +4453,18 @@ mini_emit_array_store (MonoCompile *cfg, MonoClass *klass, MonoInst **sp, gboole
 		if (sp [2]->type != STACK_OBJ)
 			return NULL;
 
+		MonoInst *index_ins = sp [1];
+#if SIZEOF_REGISTER == 8
+		if (sp [1]->type == STACK_I4) {
+			// stelemref wrapper recevies index as native int, sign extend it
+			guint32 dreg = alloc_preg (cfg);
```
Comment (vargaz) [inline src/mono/mono/mini/method-to-ir.c:4460]: Typo.

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/mini/mini-ops.h [hunk 2]
```diff
@@ -763,6 +763,7 @@ MINI_OP(OP_NEWARR, "newarr", IREG, IREG, NONE)
 MINI_OP(OP_LDLEN, "ldlen", IREG, IREG, NONE)
 /* inst_p0 is the exception name to throw or NULL */
 MINI_OP(OP_BOUNDS_CHECK, "bounds_check", NONE, IREG, IREG)
+MINI_OP(OP_BOUNDS_CHECK_SEXT, "bounds_check_sext", NONE, IREG, IREG)
```
Comment (vargaz) [inline src/mono/mono/mini/mini-ops.h:766]: Using OP_BOUNDS_CHECK for both and storing the 'sext' information in ins->inst_c0 or something could simplify some code.

Comment (BrzVlad) [reply src/mono/mono/mini/mini-ops.h]: added the flag inside `ins->backend` since both slots in `ins->data` were used by the instruction

**Pull request merged**